src/lib/Simplify.hs

```diff
@@ -835,9 +835,19 @@ projectDictMethod d i = do
 
 simplifyHof :: Emits o => Hof i -> SimplifyM i o (Atom o)
 simplifyHof hof = case hof of
-  Map f -> do
-    f' <- simplifyAtom f
-    liftM Var $ emit $ Hof $ Map f'
+  Map fun array -> do
+    (fun', Abs b recon) <- simplifyLam fun
+    array' <- simplifyAtom array
+    ans <- liftM Var $ emit $ Hof $ Map fun' array'
+    case recon of
+      IdentityRecon -> return ans
+      LamRecon reconAbs -> do
+        TabPi (TabPiType (_:>ixTy) _) <- getType array'
+        buildTabLam noHint ixTy \i -> do
+          locals <- tabApp (sink ans) $ Var i
+          ithArg <- emitAtomToName =<< (tabApp (sink array') $ Var i)
```

Comment:
Instead of emitting to name, you could `b @> SubstVal ithArg` below

Reply:
Thanks.